src/rp2040.ts

```diff
@@ -1539,7 +1539,7 @@ export class RP2040 {
       const Rm = (opcode >> 6) & 0x7;
       const Rn = (opcode >> 3) & 0x7;
       const Rt = opcode & 0x7;
-      const address = this.registers[Rm] + this.registers[Rn];
+      const address = (this.registers[Rm] | 0) + (this.registers[Rn] | 0);
```

Comment:
I'm thinking it might be better just to fix this in `readUint32()` / `writeUint32()` / `slowIO()` - always convert the address to unsigned 32-bit (with `>>> 0`) at the beginning of those functions.

This will it will fix all the other places (LDR/STR variations), and will also make sure we won't get negative numbers. I think with the current fix, writing to `0xd0000000` (SIO) will not work correctly, because it'll be converted into a negative number?

Reply:
changing to this code:

const address = this.registers[Rm] + (this.registers[Rn] | 0);

passes even this test (writing a value to DIV_UDIVIDEND)

it('should execute a `str r2, [r3, r1]` instruction where r1 + r3 > 32 bits', async () => {
    await cpu.setPC(0x20000000);
    await cpu.writeUint16(0x20000000, opcodeSTRreg(r2, r1, r3));
    await cpu.setRegisters({ r1: -4, r3: 0xd0000064, r2: 0x4201337 });
    await cpu.singleStep();
    const registers = await cpu.readRegisters();
    expect(await cpu.readUint32(0xd0000060)).toEqual(0x4201337);
    expect(registers.pc).toEqual(0x20000002);
  });

Reply:
but this solves
const address = (this.registers[Rm] + this.registers[Rn]) >>> 0;

and even this test passes:

it('should execute a `str r2, [r3, r1]` instruction where r1 + r3 > 32 bits', async () => {
    await cpu.setPC(0x20000000);
    await cpu.writeUint16(0x20000000, opcodeSTRreg(r2, r1, r3));
    await cpu.setRegisters({ r1: 0x80000000, r3: 0x50000060, r2: 0x4201337 });
    await cpu.singleStep();
    const registers = await cpu.readRegisters();
    expect(await cpu.readUint32(0xd0000060)).toEqual(0x4201337);
    expect(registers.pc).toEqual(0x20000002);
  });

Reply:
so in the end You are always right